lib/bakeAmbientOcclusion.js

```diff
@@ -0,0 +1,586 @@
+    // Copy the data over to the jimp
+    var value = 0.0;
+    jimpAO.resize(dataResolution, dataResolution);
+    for (var x = 0; x < dataResolution; x++) {
+        for (var y = 0; y < dataResolution; y++) {
+            var dataIdx = dataResolution * y + x;
+            value = 255.0 * (aoBuffer.samples[dataIdx] / aoBuffer.count[dataIdx]);
+            jimpAO.bitmap.data[dataIdx * 4 + 3] = value;
+        }
+    }
+    // Resize the data to match the goal resolution
+    jimpAO.resize(goalResolution, goalResolution, Jimp.RESIZE_BEZIER);
+}
+
+function cloneAndSetup(newMateralID, newTextureID, newImageID,
+                      oldMateralID, oldTextureID, oldImageID,
+                      materials, textures, images) {
+    var newImage;
+    var newTexture;
+    var newMaterial;
+
+    if (defined(newImageID)) {
+        var oldImage = images[oldImageID];
+        newImage = clone(oldImage);
+        newImage.extras._pipeline.jimpImage = oldImage.extras._pipeline.jimpImage.clone();
+        images[newImageID] = newImage;
+    }
+    if (defined(newTextureID)) {
+        var oldTexture = textures[oldTextureID];
+        newTexture = clone(oldTexture);
+        newTexture.source = newImageID;
+        textures[newTextureID] = newTexture;
+    } else {
+        return;
+    }
+    if (defined(newMateralID)) {
+        newMaterial = clone(materials[oldMateralID]);
+        newMaterial.texture = newTextureID;
+        materials[newMateralID] = newMaterial;
+    }
+}
+
+// Check and modify the given material to ensure every primitive gets a unique material, texture, and image
+function ensureImageUniqueness(gltf, primitive, meshPrimitiveID, state) {
+    var materialsSeen = state.materialsSeen;
+    var texturesSeen = state.texturesSeen;
+    var imagesSeen = state.imagesSeen;
+
+    var allMaterials = gltf.materials;
+    var allTextures = gltf.textures;
+    var allImages = gltf.images;
+
+    // Generate some new IDs
+    var newMaterialID = meshPrimitiveID + '_AO_material';
+    var newTextureID = meshPrimitiveID + '_AO_texture';
+    var newImageID = meshPrimitiveID + '_AO_image';
+
+    // Grab the existing material
+    var materialID = primitive.material;
+    var material = allMaterials[materialID];
+    var values = material.values;
+    var diffuse = values.diffuse;
+
+    // Check if the material has a diffuse texture material. if not,
+    // - clone the example material
+    // - clone the example texture
+    // - clone the example image. resize to resolution and set to diffuse color, if any
+    if (!defined(diffuse) || typeof diffuse !== 'string') {
+        cloneAndSetup(
+            newMaterialID, newTextureID, newImageID,
+            state.exampleMaterialID, state.exampleTextureID, state.exampleImageID,
+            allMaterials, allTextures, allImages);
+
+        var color = defaultValue(diffuse, [1.0, 1.0, 1.0, 1.0]);
+        for (var i = 0; i < 4; i++) {
```

Comment:
I'm not sure how much cleaner using Cesium `Color` would be since we would still have to load/unload each value into/out of the type. The loop was unnecessary, though, this makes more sense unrolled.